### ISSUE TITLE
sayall return player to initial state only if it is playing.

### ISSUE THREAD

**Issue by rafale77** [jishi/node-sonos-http-api#438]
Somewhat related to 436:

I have made the following observation: 
1. If some players are playing from Queue and a "sayall" is triggered, all works as expected and after the TTS command is completed, it returns to the queues.
2. If the player is on a queue and paused, the sayall will group and execute the command as expected but will not return to the initial conditions. Some players will not go back to the initial groups at random and if they do, they will remain on the TTS mp3 so a command to play will repeat the TTS ad-eternam.

I would like the players to go back to their queue even if they were paused. 436 is a bit of a workaround.

**Comment by jishi** [comment]
I have been working on improving the robustness of sayall/say, but I would need to do some extensive testing since these are intermittent issues. If you are interested, I can publish a branch with these improvements for you to evaluate?

I have made some progress, and in my environment it's much more robust, but it all depends on your specific setup.

**Comment by rafale77** [comment]
Sure I wouldn't mind testing!

**Comment by jishi** [comment]
Alright, I'll prepare and push it during the evening (CET).

**Comment by jishi** [comment]
I realized that I hadn't fixed the sayall command. I'll see if I can get it done later this week.

**Comment by rafale77** [comment]
Ok let me know. The behavior is indeed quite unpredictable on the current release.

**Comment by jhmartin** [comment]
I'd prefer it _not_ return to the queues if they are paused.

**Comment by jishi** [comment]
@jhmartin I don't think the request actually entail, forcing playback to start after a TTS announcement, merely returning the state to back where it was (queue position, etc) in order to resume it easily later on. That's at least how I picture it, and I wouldn't allow it to work any other way.

**Comment by rafale77** [comment]
@jishi. totally agree.

**Comment by jhmartin** [comment]
@jishi oh then that is ok.

**Comment by jishi** [comment]
@rafale77 I just pushed a branch called `improve-tts`, just switch branch if you are using git, or download a zip with that content here (https://github.com/jishi/node-sonos-http-api/archive/improve-tts.zip)

Let me know how it works! The main difference being it has a backup restore point based on the length of the TTS audio file, if I don't identify that the TTS playback has stopped.

**Comment by rafale77** [comment]
Got it. I will test it out. Thanks!

**Comment by rafale77** [comment]
For some reason I can't switch to this branch. I can only see the master when I do "git branch"

**Comment by jishi** [comment]
First do a `git fetch` then it should show up with `git branch --all`

**Comment by rafale77** [comment]
I did end up seeing it. Thanks.

**Comment by rafale77** [comment]
@jishi Seems a lot better but still one more thing to look at:

Result of my test: 
I have two groups both playing: group 1 plays the Queue, group 2 plays "line in"
Sayall groups all, plays the TTS, then only returns Group 1 back to playing. Group 2 which could be only one speaker, loses all status: No queue or line-in.

Running the same test with only group 2 playing, group 1 paused. Result: group 1 returns to queue paused (great!), group 2 goes to idle with no queue and no line in.

Seems to work only issue is to return the queue to all the pre-existing groups and not just one. Not sure if it is related to the fact that group 2 is playing line-in for me.

**Comment by jishi** [comment]
Hm, I can't reproduce it on my system, when any of my groups is playing line-in, it is restoring it correctly. Is this a analogue line-in, or a PLAYBAR digital line-in?

**Comment by rafale77** [comment]
It is an analog line-in. I think I did something very foolish: I started the api when the system was grouped so the api was only discovering one speaker. This might be the reason. Looking at the code, I guess I really need to start the API ungrouped for it to act on the different speakers?

**Comment by rafale77** [comment]
Nope, still not working. only the main zone recovers the state. The second zone comes back empty.

**Comment by jishi** [comment]
It doesn't matter when you start it, it is always aware of all the players and their state regardless of grouping. Since I can't reproduce this, I would need some logs. I have added more trace logging to this flow, so please update with the latest changes for that branch. 

Could you please provide trace logs for the period when you apply the sayall command? You can get trace logs by running it with the following start command (in linux and macos)
`NODE_LOG_LEVEL=trace node server.js > trace.log`

Or windows powershell: 
`$env:NODE_LOG_LEVEL='trace'`
`node server.js > trace.log` 

So start the server, then call the sayall command, wait for it to finish and trying to restore the previous state, then quit and email the logs to jimmy at shimizu.se

**Comment by rafale77** [comment]
As I was going to collect logs, after running updating and restarting the API, it appears to now work... I did not even get to the point of collecting the logs. Not sure what happened.
I have the log file but since it is working...

**Comment by jishi** [comment]
Try to gather some logs if it happens again, and get back with some feedback when you have evaluated it for some time!

**Comment by rafale77** [comment]
I will. Strange. It has been working for the different tests I have run.

**Comment by rafale77** [comment]
Seems to all work as expected. I couldn't fault it anymore. Can probably merge and close this.

**Comment by jishi** [comment]
Thanks for testing. Is now implemented in 1.2.7

**Comment by rafale77** [comment]
@jishi strange I am having the problem again once it is merged into the master. somehow would adding the log level made a difference?

**Comment by rafale77** [comment]
I think it is a timing issue. Somehow I get an error in the log saying:
"
2017-01-31T22:31:50.283Z TRACE sending M-SEARCH...
2017-01-31T22:31:50.284Z ERROR Port 5005 seems to be in use already. Make sure the sonos-http-api isn't
    already running, or that no other server uses that port. You can specify an alternative http port
    with property "port" in settings.json
"

But the API is still running and works

**Comment by jishi** [comment]
That's because you already have it running, probably a different version?
Do you have some sort of autostart that you forget about?

**Comment by rafale77** [comment]
Nope. I see that you also added the trace logs into the main. Not quite sure what is going on.
When I have 2 instances running, I usually get the error when I start the API. This only shows up in the trace log. I start the API just fine.

I ran one more test where I got out of screen and started it manually and the problem does not occur. I autostart this API with a few other things through detached screen commands. There could be a timing issue doing it this way.

What is strange is that on the branch, I restarted it many times through screen and never encountered this issue.

**Comment by jishi** [comment]
How do you autostart it? I know that pm2 can hog ports from sub-processes
even if the process is stopped.

I suggest trying to start from a clean slate. Make sure nothing is active
in port 5005, start the version you expect to test, see if it works.

I have no other suggestions for the time being...

**Comment by rafale77** [comment]
I just ran a few more tests. It is not related to autostart or screen. It seems intermittent. I am sending you some logs.
For reference I use a bash file to start a bunch of screen sessions which then start in cron but I don't think this is the problem. I just started it outside of screen and it worked once and then it failed this time If I start it within a screen session it fails to recover the status of the second zone. It does it successfully for the first one.
And yes I am sure there is nothing but this API active on port 5005.

**Comment by rafale77** [comment]
Any progress? I think adding a delay between group preset would fix it.

**Comment by jishi** [comment]
Sorry, been very busy with other things. 

We have a few options, for one, it should try and restore the biggest group first. That would minimize the phenomenon slightly. 

Adding a delay might work, but the more robust approach would be for the applyPreset to wait until the group has finished forming. It's a bit trickier, but not impossible. I'll rather try that first.

**Comment by rafale77** [comment]
This would obviously a lot more robust if it could be done. Need to detect when the preset application is completed...

**Comment by jishi** [comment]
I just pushed a change where I always restore the biggest group first. If you have the possibility to test if this improve things somewhat, that would be helpful. 

I'm gonna look into the awaiting of group forming later on. I already do a similar thing when I start the TTS announcement, since in that case I wanted it to form the group before announcing. That is not case when starting playback (in my opinion), but it could still wait with resolving the promise until the group has formed.

**Comment by rafale77** [comment]
Ok will test it shortly

**Comment by rafale77** [comment]
Seems like in my test case, running the largest of 2 groups first fixed it. Probably will run some other group size and maybe 3-4 groups to see if it totally fixes it.
